Anagrams question and answer (#170)
UWB-ACM/CTCI: 2020-01-Winter/2_string_manipulation/README.md

<!-- Don't remove -->
<a name="s2"/>

### 2. GROUP ANAGRAMS

> 

Reviewer: Provide problem source here, mostly for redundancy
```suggestion

Source: [LeetCode](https://leetcode.com/problems/group-anagrams/)

```

Two strings are anagrams if their sorted strings are equal. Then, one way
to solve this problem is by categorizing by sorting arrays.

1. Check for an empty array, if the array is not empty, then use a hashtable
where the key is the sorted string, and the value is the string that once is sorted, 
it matches the key.

2. Iterate through the array of strings. For each string in the array of strings,
create an array of chars, using the .toCharArray() function. 

3. Sort the char array, and convert it into a string. This string will be used as the
key for the key-value pair of the hashtable.

4. If the hashtable does not contains the key(string value of sorted char array), add it to the map. 

5. Add the original (unsorted) string as a new value of the key.
 
6. Return the list of list of anagrams

Time Complexity: The outer loop has complexity O(N) as we iterate through each string. 
Then, we sort each string in O(N log N) time. Therefore, the time complexity of this 
algorithm is O(N Log N)

Space Complexity: O(N)

> 

Reviewer: Missing the solution code here; add it in like so:

````suggestion

```java
public static List<List<String>> groupAnagramsOpt(String[] strs) {
    if (strs.length == 0) {
        return new ArrayList<>();
    }
    Map<String, List<String>> map = new HashMap<>();    
    for (String st:strs) {
        char[] chars = st.toCharArray();    
        Arrays.sort(chars);    
        String key = String.valueOf(chars); // Converting the sorted char sequence into string    
        if (!map.containsKey(key)) {
            map.put(key, new ArrayList<String>());
        }    
        map.get(key).add(st);
    }
    return new ArrayList(map.values());
}
```

````

#### Optimal Solution

Two strings are anagrams only if the characters counts are the same. 

Another way to solve this problem is by transforming each string into a character count 
"count", consisting of 26 non-negative integers representing the 
number of times each character occurs in each string. 

1. Check for an empty array, if the array is not empty, then use a hashtable
where the key will be represented by a string delimited with '#' characters. For 

> example abbccdddd will be represented as "1#2#2#4" where 1,2, and 4 are the number of 

Reviewer: ```suggestion
example abbccdddd will be represented as "1#2#2#4#0...#0" where 1,2, and 4 are the number of 
```

> times each character occurs in the string.

Reviewer: This wasn't totally clear until I read through the LeetCode problem statement again. Chars which aren't present should be denoted with a 0 placeholder.

2. Create an integer array of 26 spaces 'count'. Each index in the array represent a letter in the 
alphabet ( a = 0, b = 1...)

3. Iterate through the string array. For each string in the array, add 1 everytime a character 
occurs in the string in the correspoding index (index representing character).

4. Use a string builder to iterate  through 'counts'. Append ("#") before appending the total
counts (times) per character.

5. Convert the string builder into a string and add it to the map. The value for this key is the
current original string.

Time Complexity: O(NK). Counting every string in the array is O(N), and counting each character
in each string is O(K), where K is the max length of a string in the string array.

Space Complexity: O(N)

#### Testing The Solutions OR Driver For Solution

> [Available here: ](./2_string_manipulation/ Anagrams/ Anagram.java)

Reviewer: The link text should be identical to the filepath for the .java file. Also, personal nitpick, but spaces in directory or file names are discouraged.

```suggestion
[Available in the repository.](./2_string_manipulation/2. Anagrams/Anagram.java)
```
